### PR TITLE
Laboratory work #1, Julia Lisovskaia

### PR DESCRIPTION
**Pull request by JuliaLisovskaia**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+print(a)
+print(b)
+
+
+a= int ('2') # cast/casting
+b=float ('0.5')
+
+print(a,b)
+
+print('Арифметические операции')
```
Comment (demid5111) [inline main.py]: never use russian language in Python scripts

Comment (JuliaLisovskaia) [reply main.py]: Этот комментарий я считаю избыточным

File: main.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
+signal=[8.4,3.3,4.5,1.2,5.6,8.1,4.0,2.3,4.1]
```
Comment (demid5111) [inline main.py]: remove from git

File: lab_2/main.py [hunk 3]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
 
 # Lab 2 implementation goes below
 
+# def signal_check(signal):
+#     if type(signal) != list or len(signal) == 0:
+#         return None
+#     for s in signal:
+#         if type(s) != float:
+#             return None
+
```
Comment (demid5111) [inline lab_2/main.py:51]: ### Merging changes via merge commit

`git fetch --all`

`git merge origin/main`

Conflicts to be resolved

### Merging changes via rebase (default)

`git pull --rebase`

`git rebase --abort` - to stop rebasing

File: lab_2/main.py [hunk 4]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+
 def calculate_sdnn(signal: list):
     """Calculating SDNN for RR-intervals from the list"""
-    pass
+    # signal_check(signal)
```
Comment (demid5111) [inline lab_2/main.py]: rework to accept common checks

File: lab_2/main.py [hunk 5]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+
+    if type(signal) != list or len(signal) == 0:
+        return None
+    for s in signal:
+        if type(s) != float:
+            return None
+
+    rr_sum = 0
```
Comment (demid5111) [inline lab_2/main.py]: use built-in function for calculating sum of objects in a list

File: lab_2/main.py [hunk 6]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+
+    if type(signal) != list or len(signal) == 0:
+        return None
+    for s in signal:
+        if type(s) != float:
+            return None
+
+    rrd_sq_sum = 0
+    for i in range(len(signal)):
+        if i < len(signal) - 1:
```
Comment (demid5111) [inline lab_2/main.py]: condition should be implicitly included in the range that is used for the loop

File: lab_2/main.py [hunk 7]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+
+    rr_sum = 0
+    for rr in signal:
+        rr_sum += rr
+    rr_mean = rr_sum / len(signal)
+    sum_f = 0
+    for rr in signal:
+        sum_f += (rr - rr_mean) ** 2
+    sdnn_raw = (sum_f / len(signal)) ** (1/2)
+    sdnn = round(sdnn_raw, 2)
```
Comment (demid5111) [inline lab_2/main.py]: intermediate variables

File: lab_2/main.py [hunk 8]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
-    pass
+    # signal_check(signal)
+
+    if type(signal) != list or len(signal) == 0:
+        return None
+    for s in signal:
+        if type(s) != float:
+            return None
+
+    rrd_sum = 0
```
Comment (demid5111) [inline lab_2/main.py]: create a separate function for extraction of differences

File: lab_2/main.py [hunk 9]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+    for s in signal:
+        if type(s) != float:
+            return None
+
+    if type(threshold) != float and type(threshold) != int:
+        return None
+    nn = 0
+    for i in range(len(signal)):
+        if i < len(signal) - 1:
+            if abs(signal[i] - signal[i + 1]) <= threshold:
```
Comment (demid5111) [inline lab_2/main.py]: reuse rrd calculation

File: lab_2/main.py [hunk 10]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
 
 
 def save_hrv_in_file(hrv_characteristics: dict, path: str):
     """Calculating HRV time-scale metrics for RR-intervals from the list"""
-    pass
-
+    if type(hrv_characteristics) != dict or type(path) != str:
+        return -1
+    with open(path, 'w') as file:
+        for key, value in hrv_characteristics.items():
```
Comment (demid5111) [inline lab_2/main.py]: single write operation is required

File: lab_2/main.py [hunk 11]
```diff
@@ -30,30 +30,107 @@ def read_ecg_rr_file(file_path: Path):
...
+            file.write(f'{key}\t{value}\n')
```
Comment (demid5111) [inline lab_2/main.py]: no new line at the end of the file

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**